File: backend/homomorphic_aggregator.py
#!/usr/bin/env python3
"""
Homomorphic Aggregator — Radix Slot Packing Scheme
===================================================

Each vote is a SINGLE Paillier ciphertext encrypting:

    plaintext = B ^ candidateIndex

where B (vote_block) is the smallest power of 10 strictly greater than the
total number of registered voters for the election.  This prevents carry-bleed
between candidate slots while minimising slot width, maximising the number of
candidates that fit within the Paillier modulus.

Aggregation (homomorphic multiplication / Paillier addition):

    E(v_1) * E(v_2) * ... = E(v_1 + v_2 + ...) = E(msum)

Only ONE decryption is performed at Phase 4.

Extraction at Phase 4:

    count[i] = floor(msum / B^i) mod B

Voter identity is NOT embedded — auditability is provided by the on-chain
EncryptedVoteCast event that maps voter address → IPFS CID.

Usage:
    python homomorphic_aggregator.py <votes_json_file>

Input JSON:
    {
        "public_key_n"   : "...",
        "num_candidates" : 3,
        "vote_block"     : "1000",       <-- B from compute_vote_block(total_voters)
        "encrypted_votes": ["ct1", "ct2", ...]
    }

Output JSON:
    {
        "encrypted_total" : "big ciphertext string",
        "vote_count"      : 3,
        "num_candidates"  : 3,
        "vote_block"      : "1000",
        "method"          : "Paillier-RadixPack"
    }
"""

import sys
import json
from paillier_crypto import PaillierPublicKey



def perform_homomorphic_aggregation(public_key_n, encrypted_votes):
    """
    Multiply all ciphertexts to obtain E(msum).

    Paillier homomorphic property:
        E(a) * E(b) mod n^2  =  E(a + b) mod n^2

    Args:
        public_key_n (str):         Paillier modulus as decimal string.
        encrypted_votes (list[str]): Ciphertext strings (one per voter).

    Returns:
        str: The encrypted aggregate sum as a decimal string.
    """
    pk = PaillierPublicKey(int(public_key_n))
    n_sq = pk.n_squared

    encrypted_total = 1  # Paillier identity: E(0) = 1
    for ct_str in encrypted_votes:
        encrypted_total = (encrypted_total * int(ct_str)) % n_sq

    return str(encrypted_total)


def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: python homomorphic_aggregator.py <votes_json_file>"}))
        sys.exit(1)

    try:
        with open(sys.argv[1], 'r') as f:
            data = json.load(f)

        public_key_n    = data["public_key_n"]
        num_candidates  = int(data["num_candidates"])
        vote_block      = int(data["vote_block"])      # B from compute_vote_block
        encrypted_votes = data["encrypted_votes"]      # list of ciphertext strings

        if not encrypted_votes:
            print(json.dumps({"error": "No encrypted votes provided"}))
            sys.exit(1)

        encrypted_total = perform_homomorphic_aggregation(public_key_n, encrypted_votes)

        print(json.dumps({
            "encrypted_total" : encrypted_total,
            "vote_count"      : len(encrypted_votes),
            "num_candidates"  : num_candidates,
            "vote_block"      : str(vote_block),
            "method"          : "Paillier-RadixPack"
        }))

    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)


if __name__ == "__main__":
    main()
if __name__ == "__main__":
    main()

File: dapp/src/utils/homomorphicAggregator.js
/**
 * Homomorphic Aggregation for Paillier Encrypted Votes (Browser)
 * 
 * Performs homomorphic addition on Paillier-encrypted votes.
 * Multiplies ciphertexts to compute the encrypted sum without decryption.
 * 
 * Paillier homomorphic property:
 *   E(m1) × E(m2) = E(m1 + m2) mod n²
 */

/**
 * Perform modular exponentiation: (base^exp) mod modulus
 * Uses BigInt for large number arithmetic
 */
function modPow(base, exp, modulus) {
  if (modulus === 1n) return 0n;
  
  let result = 1n;
  base = base % modulus;
  
  while (exp > 0n) {
    if (exp % 2n === 1n) {
      result = (result * base) % modulus;
    }
    exp = exp >> 1n;
    base = (base * base) % modulus;
  }
  
  return result;
}

/**
 * Perform modular multiplication
 */
function modMul(a, b, modulus) {
  return (BigInt(a) * BigInt(b)) % BigInt(modulus);
}

/**
 * Perform homomorphic addition on encrypted votes
 * 
 * @param {string} publicKeyN - The public key modulus N (as string)
 * @param {Array<string>} encryptedVotes - Array of encrypted votes (ciphertexts as strings)
 * @returns {string} - The encrypted sum (as string)
 */
export function performHomomorphicAddition(publicKeyN, encryptedVotes) {
  if (!publicKeyN || encryptedVotes.length === 0) {
    throw new Error('Invalid input: publicKeyN and encryptedVotes required');
  }

  // Convert to BigInt
  const n = BigInt(publicKeyN);
  const nSquared = n * n;
  
  // Start with identity element: E(0) = 1 (mod n²)
  let encryptedSum = 1n;
  
  // Multiply all ciphertexts (homomorphic addition)
  for (const encryptedVote of encryptedVotes) {
    const ciphertext = BigInt(encryptedVote);
    encryptedSum = modMul(encryptedSum, ciphertext, nSquared);
  }
  
  return encryptedSum.toString();
}

/**
 * Aggregate votes by candidate index
 * Groups votes by candidate and computes encrypted sums per candidate
 * 
 * @param {string} publicKeyN - The public key modulus N
 * @param {Array<Object>} votes - Array of vote objects with {candidateIndex, ciphertext}
 * @returns {Object} - Object mapping candidateIndex to encrypted sum
 */
export function aggregateVotesByCandidate(publicKeyN, votes) {
  // Group votes by candidate index
  const votesByCandidate = {};
  
  for (const vote of votes) {
    const idx = vote.candidateIndex;
    if (!votesByCandidate[idx]) {
      votesByCandidate[idx] = [];
    }
    votesByCandidate[idx].push(vote.ciphertext);
  }
  
  // Perform homomorphic addition for each candidate
  const encryptedTotals = {};
  
  for (const [candidateIndex, ciphertexts] of Object.entries(votesByCandidate)) {
    encryptedTotals[candidateIndex] = performHomomorphicAddition(publicKeyN, ciphertexts);
  }
  
  return encryptedTotals;
}

/**
 * Create aggregation result object
 * 
 * @param {string} encryptedTotal - The encrypted total
 * @param {number} voteCount - Number of votes aggregated
 * @returns {Object} - Result object
 */
export function createAggregationResult(encryptedTotal, voteCount) {
  return {
    encrypted_total: encryptedTotal,
    vote_count: voteCount,
    method: 'Paillier Homomorphic Addition',
    timestamp: new Date().toISOString()
  };
}

export default {
  performHomomorphicAddition,
  aggregateVotesByCandidate,
  createAggregationResult
};
